Reworking roadmap to have short-term high level vision, linking to google doc for details (#5827)
sourcegraph/sourcegraph: doc/dev/roadmap/index.md

- **For developers:** the best way to answer questions and get unblocked while writing, reviewing, or reading code.
- **For organizations** (engineering leaders and internal tools teams): the infrastructure for developer tools and data.

> This roadmap is an overview of what’s coming in the next 3-6 months. The project details for how we get there are in our [Project Roadmap Google Doc](https://docs.google.com/document/d/1cBsE9801DcBF9chZyMnxRdolqM_1c2pPyGQz15QAvYI/edit?usp=sharing). Our high-level vision is outlined in the [Sourcegraph master plan](https://about.sourcegraph.com/plan).

Reviewer: ```suggestion
This roadmap is an overview of what’s coming in the next 3-6 months.  More details are in the [project roadmap](https://docs.google.com/document/d/1cBsE9801DcBF9chZyMnxRdolqM_1c2pPyGQz15QAvYI/edit?usp=sharing). Our high-level vision is outlined in the [Sourcegraph master plan](https://about.sourcegraph.com/plan).
```

- Use sentence case not title case
- No need to say "Google Doc"
- Remove unneeded words

We ship a release on the [20th day of each month](../releases.md#releases-are-monthly).

Want to help us achieve these goals? [We're hiring!](https://github.com/sourcegraph/careers/blob/master/job-descriptions/software-engineer.md)

## Overview

We're continually improving Sourcegraph's core features for developers:

- Code search and navigation (with code intelligence)
- Integration into code review
- Automation of large-scale code changes

> The current product focus is on the usability, scalability, and automation features of the product. We strive to make Sourcegraph intuitive to use for anyone looking for answers about their code, and easy for admins to configure and maintain. We expect Sourcegraph to perform reliably at scale for our largest customers. We intend to create code change automation features that organically spread value throughout an organization.

Reviewer: - Use bullet points for readability. (See below.)
- Personally I find it weird when someone says they are focused on more than one thing. Humans can only focus on one thing, so it sounds unfocused. I would just speak in terms of priorities.
- "organically spread value throughout an organization" sounds vague. If you use the bullets below, I suggest something like "Automation: clean up tech debt and make other large-scale code changes across your entire codebase".

---

Our current product priorities are:

- Usability: ...
- Scalability: ...
- Automation: ...

Author: Updated.

## In 3-6 months, Sourcegraph will have

### Improved usability

Sourcegraph is intuitive to use for a wide range of roles, from developers to PMs, engineering managers, data analysts, and more. It adds value to this wide range of roles by making more information about code available, such as language statistics and dependency graphs. [Search has an improved UI](https://docs.google.com/document/d/1Vo7HlwO_HgrK8O-VEIZ9wHuSyHdEA0zk9qucNCoF0jg/edit?usp=sharing) that makes it more accessible, easier to use, and faster to drill down on what you're looking for.

> Improving upon on our out-of-the-box code intelligence, Sourcegraph provides fast and precise code intelligence in the browser witha a quick and easy setup to your repository.

Reviewer: - fix typo witha -> with a
- "out-of-the-box code intelligence" is introducing a new term "out-of-the-box" that we haven't used yet. Be consistent and call it "basic", and link to a code intelligence doc page that describes this (if one doesn't exist, then you should make a quick one).
- The key differentiator between basic and precise code intel is that the latter is precise, but the first adjective you use is "fast" ("fast and precise"). This is confusing.
- "provides x with a setup to your repository" sounds wrong grammatically.

This bit is really important to make clear because the question of "what languages does Sourcegraph support?" is such a common one for our customers. Can you update this given the feedback above?

Author: Updated, let me know what you think!

### Enhanced scalablity, reliability, and security

> Sourcegraph performs at the scale of our largest customers under a wide variety of configurations and deployment infrastructure. [Code search is fast at large scale](https://docs.google.com/document/d/18w8T_KzYxQye8wg1g01QpMOX4_ERTtbOxMBRYaOEkmk/edit?usp=sharing) (~80k repositories), and the API can support enterprise level usage. Sourcegraph enforces repository permissions using ACLs from Bitbucket Server, GitHub, and GitLab.

Reviewer: gdoc has 100k, this says 80k. Is that because 80k is the "next few months" goal and 100k is the longer-term vision goal?

Author: Yes, but for consistency's sake, I'll use 80k and when we want to push higher I'll update all the numbers!

> Admins understand the health of their instances and are alerted proactively if things fail. It is clear what repositories are configured, and what their syncing status is.

Reviewer: "are configured": This is another area where being consistent about language is important. To us, it's clear that a repository being configured means it's searchable and browseable on Sourcegraph. I'm not sure that's clear to customers. Being configured could mean they've set up code intelligence for their repository, or maybe it confuses them into thinking Sourcegraph needs some other kind of per-repository configuration.

"are available" or "are mirrored on Sourcegraph" is better.

Also, I think we should avoid using the term "syncing" because it is ambiguous about whether it's 1-way or 2-way. 2-way-sync introduces risks and fear into our customers (eg that Sourcegraph might clobber their repo accidentally).

Author: While trying to refactor, I realized this was too specific to add value - what I was trying to capture is the work around the status indicator, but I think that is clear enough in the project roadmap and isn't needed here.

> ### Large scale code automation

Reviewer: ```suggestion
### Large-scale code change automation
```

to be consistent

> You can use Sourcegraph to [automate large-scale code changes](https://about.sourcegraph.com/product/automation) to remove legacy code, fix critical security issues, and pay down tech debt. Developers can collaborate on campaigns across thousands of repositories, and share, notify, and alert accordingly to get these changes pushed through.

Reviewer: ```suggestion
Use Sourcegraph to [automate large-scale code changes](https://about.sourcegraph.com/product/automation) to remove legacy code, fix critical security issues, and pay down tech debt. You can create campaigns across thousands of repositories and code owners. Sourcegraph automatically creates and updates all of the branches and pull requests, and you can track progress and activity in one place.
```

- **For developers:** the best way to answer questions and get unblocked while writing, reviewing, or reading code.
- **For organizations** (engineering leaders and internal tools teams): the infrastructure for developer tools and data.

This roadmap is an overview of what’s coming in the next 3-6 months. The project details for how we get there are in our [Project Roadmap Google Doc](https://docs.google.com/document/d/1cBsE9801DcBF9chZyMnxRdolqM_1c2pPyGQz15QAvYI/edit?usp=sharing). Our high-level vision is outlined in the [Sourcegraph master plan](https://about.sourcegraph.com/plan).

> We ship a release on the [20th day of each month](http://localhost:5080/dev/releases#releases-are-monthly).

Reviewer 2: I don't think you meant to commit this?

Author: Nope, that was a mistake from testing locally, thanks for the catch!

> **Next release:** [Sourcegraph 3.9](https://github.com/sourcegraph/sourcegraph/issues?q=is%3Aopen+is%3Aissue+milestone%3A3.9+label%3Aroadmap) (ships on October 20, 2019).

Reviewer 2: I think we can just delete this so we don't have to update it every month.

```suggestion
```

Want to help us achieve these goals? [We're hiring!](https://github.com/sourcegraph/careers/blob/master/job-descriptions/software-engineer.md)

## Overview

We’re continually improving Sourcegraph’s core features for developers:

- Code search and navigation (with code intelligence)
- Integration into code review
- Automation of large-scale code changes

The current product focus is on the usability, scalability, and virality of the product. We strive to make Sourcegraph intuitive to use for anyone looking for answers about their code, and easy for admins to configure and maintain. We expect Sourcegraph to perform reliably at scale for our largest customers. We intend to create viral features that organically spread value throughout an organization.

## In 3-6 months, Sourcegraph will be

> ### User-friendly

Reviewer 2: There is an implication here that we aren't user friendly today. Is there a way to frame this without implying that we aren't user friendly today but are planning to make improvements in this direction in the future?

Author: Yeah, I was worried about this too. Other iterations had the titles as:
- Usability
- Scalability
- Virality

but it didn't flow as well and the titles felt misplaced. What about changing the above h2 to be "## In 3-6 months, Sourcegraph will have improved" and then use the 3 titles above?

Reviewer 2: I like the changes you made in 037f5ab37d562463b42cf35387bd4e789be88c09

Improving upon on our out-of-the-box code intelligence, Sourcegraph provides fast and precise code intelligence in the browser witha a quick and easy setup to your repository.

### Scalable, reliable, and secure

Sourcegraph performs at the scale of our largest customers under a wide variety of configurations and deployment infrastructure. [Code search is fast at large scale](https://docs.google.com/document/d/18w8T_KzYxQye8wg1g01QpMOX4_ERTtbOxMBRYaOEkmk/edit?usp=sharing) (~80k repositories), and the API can support enterprise level usage. Sourcegraph enforces repository permissions using ACLs from Bitbucket Server, GitHub, and GitLab.

Admins understand the health of their instances and are alerted proactively if things fail. It is clear what repositories are configured, and what their syncing status is.

> ### Viral

Reviewer 2: "viral" doesn't seem like the best way to frame automation. It makes it sound like something that benefits us and not our customers. I would want to avoid using that word anywhere in this doc.

Author: I was worried about that too - my intention here was to indicate that we are going to spread value in a faster way at an organization. The virality is intended to help the customer by making the product more useful to more people. Pushing an update.

Reviewer 2: The updated language in 037f5ab37d562463b42cf35387bd4e789be88c09 looks good to me
